### DISCUSSION TITLE
Migration Question

### DISCUSSION THREAD

**Discussion by MaxiStefan** [simolus3/drift#2742] [Q&A] [answered]
Hi!

I migrated the database to a new version by removing a column from a table. I generated the new schema and I wanted to do a migration test. I got the following error for another table and totally unrelated.

```
project_users:
 columns:
  created_date:
   Not equal: `NOT NULL DEFAULT '2023-11-22T14:21:41.095772Z'` (expected) and `NOT NULL DEFAULT '2023-11-22T14:21:41.009747Z'` (actual)
  last_updated:
   Not equal: `NOT NULL DEFAULT '2023-11-22T14:21:41.095772Z'` (expected) and `NOT NULL DEFAULT '2023-11-22T14:21:41.009747Z'` (actual) 
```

This is how the table is defined
```

DateTimeColumn get createdDate =>
    dateTime().withDefault(Constant(DateTime.now().toUtc()))();
DateTimeColumn get lastUpdated =>
    dateTime().withDefault(Constant(DateTime.now().toUtc()))();
```

I understand the concept behind withDefault / clientDefault. But why does this come up as a problem for the schema verification? Any way to overcome this error?

Thanks in advance! :)

**Comment by simolus3** [comment]
> But why does this come up as a problem for the schema verification?

Because the schema verification creates all tables again and you have `DateTime.now()` in the table's definition. So the schema of the table depends on when the table instance is created. So if the verification starts 80ms after you've created the database for the first time, this is what you'll see.

> Any way to overcome this error?

If you really want the default value to be the timestamp of when the table / that column has been created in the database, store `DateTime.now().toUtc()` in a static field and reference that in the `Constant`.
If you want the default value to be the time of the insert, use `withDefault(currentDateAndTime)`, which uses a SQL expression to obtain the current time. `clientDefault(() => DateTime.now().toUtc())` would be an alternative as well.

**Reply by MaxiStefan** [reply]
How do I go about migrating this table with changing the DEFAULT value? I understand that alter table does not support this? I've seen some workarounds online but they look like a lot of work. I would not even mind completely removing a DEFAULT value.

I tried to remove the .withDefault part and ran the build_runner & the schema generations but the error turned from NOT NULL (expected) to 2023-11-22T14:21:41.009747Z (actual)

**Reply by simolus3** [reply, marked as answer]
> from NOT NULL (expected) to 2023-11-22T14:21:41.009747Z (actual)

That's probably due to the missing schema migration, right?

> How do I go about migrating this table with changing the DEFAULT value? I understand that alter table does not support this? 

The recommended workaround is [this 12-step procedure](https://www.sqlite.org/lang_altertable.html#otheralter), which can be done with drift in a single call: [Migrator.alterTable](https://drift.simonbinder.eu/api/drift/migrator/altertable). If you just want to change column constraints but leave data unchanged, you can use [the default options](https://drift.simonbinder.eu/docs/migrations/api/#changing-column-constraints).

**Reply by MaxiStefan** [reply]
It worked, thanks!
